coordinator/src/notifications.rs

```diff
@@ -0,0 +1,355 @@
+use crate::db::positions;
+use crate::db::user;
+use crate::position::models::Position;
+use anyhow::Context;
+use anyhow::Result;
+use diesel::PgConnection;
+use std::fmt::Display;
+use time::OffsetDateTime;
+use tokio::sync::mpsc;
+
+/// Consider a position expiring soon if it expires in less than this time
+const START_OF_EXPIRING_POSITION: time::Duration = time::Duration::hours(1);
+
+/// Consider a position expired if it expired more than this time ago
+const END_OF_EXPIRED_POSITION: time::Duration = time::Duration::hours(1);
+
+/// Types of notification that can be sent to 10101 app users
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NotificationKind {
+    /// Coordinator would like to settle the channel
+    ChannelClose,
+    PositionSoonToExpire,
+    PositionExpired,
+}
+
+impl Display for NotificationKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NotificationKind::ChannelClose => write!(f, "ChannelClose"),
+            NotificationKind::PositionSoonToExpire => write!(f, "PositionSoonToExpire"),
+            NotificationKind::PositionExpired => write!(f, "PositionExpired"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Notification {
+    pub user_fcm_token: FcmToken,
+    pub notification_kind: NotificationKind,
+}
+
+impl Notification {
+    pub fn new(user_fcm_token: FcmToken, notification_kind: NotificationKind) -> Self {
+        Self {
+            notification_kind,
+            user_fcm_token,
+        }
+    }
+}
+
+/// Actor managing the notifications
+pub struct NotificationService {
+    notification_sender: mpsc::Sender<Notification>,
+}
+
+impl NotificationService {
+    /// Start the notification service
+    ///
+    /// If an empty string is passed in the constructor, the service will not send any notification.
+    /// It will only log the notification that it would have sent.
+    pub fn new(fcm_api_key: String) -> Self {
+        if fcm_api_key.is_empty() {
+            // Log it as error, as in production it should always be set
+            tracing::error!("FCM API key is empty. No notifications will not be sent.");
+        }
+
+        let (notification_sender, mut notification_receiver) = mpsc::channel(100);
+
+        // TODO: use RAII here
+        tokio::spawn({
```

Comment:
🔧 I don't think we need that task at all if `fcm_api_key.is_empty`. You could move the `fcm_api_key.is_empty` condition outside of this function and only create a notification service if the `fcm_api_key` is set.

Reply:
the reason I kept it (as stipulated in the previous commit) is to ensure we can at least see the correct behaviour in the logs, whilst running locally etc.